Lib/unittest/case.py

```diff
@@ -63,24 +64,65 @@ def testPartExecutor(self, test_case, subTest=False):
         except _ShouldStop:
             pass
         except:
-            exc_info = sys.exc_info()
             if self.expecting_failure:
-                self.expectedFailure = exc_info
+                self.expectedFailure = sys.exc_info()
             else:
                 self.success = False
                 if subTest:
-                    self.result.addSubTest(test_case.test_case, test_case, exc_info)
+                    self.result.addSubTest(test_case.test_case, test_case, sys.exc_info())
                 else:
-                    _addError(self.result, test_case, exc_info)
-            # explicitly break a reference cycle:
-            # exc_info -> frame -> exc_info
-            exc_info = None
+                    _addError(self.result, test_case, sys.exc_info())
+                if self.debug:
+                    _handle_debug_exception(
+                        self.debug,
+                        on_crash=not subTest and
+                        (lambda: self.pm_frame_holds.append(_AutoDelRunner(self.pm_cleanup))))
         else:
             if subTest and self.success:
                 self.result.addSubTest(test_case.test_case, test_case, None)
         finally:
             self.success = self.success and old_success
 
+class _AutoDelRunner(object):
+    def __init__(self, func):
+        self.func = func
+    def __del__(self):
+        self.func()
```

Comment:
How robust is it to rely on `__del__` for this?  (as it’s not guaranteed when an implementation calls `__del__` + what global names are still available at tha time)

Reply:
(Updated the above sample code for trying such things out.)

It relys on the traceback being freed before the module globals are wrecked (`Py_EndInterpreter / Py_FinalizeEx`). And ideally the outer runner stub may do a `gc.collect()` or equivalent before exiting - which is generally a good practice.

This AutoDelRunner for delayed teardown after crash is used in `--debug` mode only, thus after an outer runner de-refs the traceback.  Not in the other in-line modes.
Note: The existing `TestCase.debug()/TestSuite.debug()` method doesn't care about delayed teardowns. Its principally a rather harsh and invasive mode. But it enables a seamless no-callback postmortem, like normal script execution - which has a high value in various environments.

With refcounted CPython this runs reliable when the traceback and its frames are refcounted to zero **before** the loaded modules are wrecked in kind of random order in `Py_EndInterpreter / Py_FinalizeEx`.  The `<_AutoDelRunner>` and its `func` are just linearly linked from the frames in TestCase.run()/TestSuite.run()

I tried it a lot: When there is no extra ref cycle holding the test code frames (inside the tested code) this happens immediately (without `gc.collect()`) as soon as the last ref to the tb goes - e.g. when `sys.last_traceback & sys.last_value.__traceback__` are deleted or the outer exception handler code (of a runner stub of an IDE, interactive, super-runner ..  )  lets go. 

I didn't look up so far if `Py_EndInterpreter / Py_FinalizeEx` destroys `sys.last_traceback & sys.last_value` at priority before wrecking the module globals - but experimentally that seems to happen...  (E.g. in the above sample code, when run without the `atexit` and directly from shell without external runner stub, then still  `"Del X"` always comes before `"-- delframe TDC"`  - after I did those fixes regarding some ref cycle breaks inside unittest. )  But the `--debug` mode is typically used via a runner stub handling the exception anyway.

Otherwise (ref cycles inside test code or non-refcounted implementations)  it happens when a full `gc.collect()` is run, which a good runner stub should generally do before exiting in such cases.

In non-refcounted implementations (PyPy;  Jython and IronPython are rather dead I think?) a full `gc.collect()` or equivalent is more important to do for the runner stub.

So far this does not require an external runner to call a special `delayed_teardown` function, or to even know that a  unittest program is run. Principally such a function could be offered for extreme cases (unbreakable ref cycle (`gc.garbage`) within the test code). But I think this would be hyperbole and rather unpractical. 
After all its about a debug mode during dev in the case of non-trivial errors - and potentially a leftover temporary file or so. The test run doesn't continue otherwise.

Reply:
With the above sample run with `--debug` directly from shell and the line `##gc.collect()` activated one could try to construct some breaking test code which would manage to block the delayed teardown:  Constructing some `gc.garbage` via certain cyclically linked classes having a `__del__` method and the cycle referencing back to the test code frame...
(A smarter runner stub could then go about dismantling gc.garbage ...)